### ISSUE TITLE
Warning: date_timezone_set() expects parameter 1 to be DateTime, boolean given

### ISSUE THREAD

**Issue by AlexShapka** [backdrop/backdrop-issues#4132]
**Description of the bug**
At certain conditions, the https://github.com/backdrop/backdrop/blob/2a719a4d1f5ce20c47092c8fbf6ccf1cf81f15fa/core/includes/common.inc#L2394 is causing the error in title.

**Steps To Reproduce**
To reproduce the behavior:
1. Download and enable the Demo contrib module.
2. Got to `/admin/structure/demo` page and you will see the error.

**Additional information**

As I explained on [my PR](https://github.com/backdrop/backdrop/pull/2934), I myself am inclined that this PR is not accepted, because the same error that is discussed for Drupal on https://www.drupal.org/node/1874768 is always caused by some contributed module which feeds incorrect parameter to `format_date()` function and that in turn breaks the core. The error (backdrop-contrib/demo#1) caused by the demo module is an example of this. So ideally this error should always be addressed in the responsible contrib module.

However, reading https://www.drupal.org/node/1874768 and seeing that all kinds of bad coding in conrib modules can cause this, I thought maybe safeguarding the core files and make them foolproof is a good idea.

**Comment by stpaultim** [comment]
> I myself am inclined that this PR is not accepted

@AlexShapka - Just to clarify the language here - I understand this to mean that you are not sure if this PR is a good idea and understand if it is not accepted. Is that correct?

**Comment by AlexShapka** [comment]
Well, generally I don't like messing with core when cause of the issue is in contrib module, so I am totally fine if the proposed patch is refused. However, remembering how easy Drupal core is breakable by all kind of stupid code in the contribs, I thought maybe making Backdrop core foolproof could be a good idea. I am not sure how flexible or conservative Backdrop community is about changing core, so it's totally to the discretion of Backdrop core maintainers.

**Comment by AlexShapka** [comment]
The more I sit on this, the more I am convinced that instead of trying to make core foolproof, we should expect the offensive contrib module to correct its bad input sent to the `format_date()` function. Because, hiding the actual problem and making impossible to backtrace the error could create even worse situations. So I am recalling my own PR. Let' concentrate on https://github.com/backdrop-contrib/demo/issues/1 instead.

**Comment by klonos** [comment]
> I am not sure how flexible or conservative Backdrop community is about changing core, so it's totally to the discretion of Backdrop core maintainers.

@AlexShapka as explained on Gitter, I think that we should reopen this, and also the PR that you have created.

Ideally, I would like us to introduce a new helper function in Backdrop core something like this perhaps:

```
function is_valid_timestamp($timestamp) {
  try {
    // Prepending '@' as the $time parameter denotes UNIX timestamp.
    new DateTime('@' . $timestamp);
  }
  catch(Exception $e) {
    return FALSE;
  }
  return TRUE;
}
```

**Comment by AlexShapka** [comment]
Nice idea. Can you elaborate how returning TRUR or FALSE from the proposed function should change the following lines:

```
  // Create a DateTime object from the timestamp.
  $date_time = date_create('@' . $timestamp);

  // Set the time zone for the DateTime object.
  date_timezone_set($date_time, $timezones[$timezone]);
```

The `date_create()` function will kick in regardless of the return of the proposed function and it breaks if it is FALSE anyway.

**Comment by AlexShapka** [comment]
Actually, I see how it should work: the offending contrib module should still check if it is returning the correct input with `is_valid_timestamp($reset_date)` and the above lines in `common.inc` stay unchanged intact with just new function introduced in the core.

But then the initial problem of letting contrib modules have a bad code is still there. I mean with the new function introduced the core still affected by bad input coming from contrib.

**Comment by klonos** [comment]
> Actually, I see how it should work...

Yup, precisely what I was thinking. Remains to be validated by a core committer, as I am not a very experienced developer.

> But then the initial problem of letting contrib modules have a bad code is still there. ...

You are right. To that extend, the solution that will come from https://www.drupal.org/node/1874768 should help improve things.

**Comment by AlexShapka** [comment]
> You are right. To that extend, the solution that will come from https://www.drupal.org/node/1874768 should help improve things.

I am not sure, because I tested the proposed patch there and it just makes the error notice go away without giving any hint to developer or logs that bad input was sent to the core. So as you suggested elsewhere (probably on gitter) every instance of incorrect data coming from contrrib modules should be at least logged, so devs could backtrace and troubleshoot them.

**Comment by klonos** [comment]
...yup. Looks like we may need to enhance the D7 solution to include a `watchdog()` call, with a useful message.

**Comment by AlexShapka** [comment]
Ok, then, let's open a new PR, but with better code suggestion this time. Maybe something like this:

```
function is_valid_timestamp($timestamp) {
  try {
    // Prepending '@' as the $time parameter denotes UNIX timestamp.
    new DateTime('@' . $timestamp);
  }
  catch(Exception $e) {
    new watchdog() entry here
    return FALSE;
  }
  return TRUE;
}
```

What about make the watchdog message even better to include backtrace information and implement something like https://api.drupal.org/comment/32338#comment-32338?

Also where do you think the proposed function should be? In the same `common.inc` file or elsewhere?

**Comment by klonos** [comment]
Yes, backtrace info would be very useful 👍 ...as to where this function should live or what it should be called, let's just go with `common.inc` and `is_valid_timestamp()` and wait for feedback from core maintainers and others that will review it.

**Comment by AlexShapka** [comment]
Is it ok if we just continue to discuss possible changes without re-creating the deleted PR until the core maintainers say their stance on this?

And in case, if the proposed function will be introduced to core to catch errors, then:

> How about:
> 
> ```
> $reset_date = config_get('demo.settings', 'demo_reset_last');
>   // @todo: change this check to use is_valid_timestamp().
>   // See https://github.com/backdrop/backdrop-issues/issues/4132
>   try {
>     // Prepending '@' as the $time parameter denotes UNIX timestamp.
>     $reset_date = new DateTime('@' . $reset_date);
>   }
>   catch(Exception $e) {
>     $message = t("<strong>Error:</strong> @message <br>", array('@message' => $e->getMessage()));
>     $message .= t("<strong>File:</strong> @file <br>", array('@file' => $e->getFile()));
>     $message .= t("<strong>Line:</strong> @line", array('@line' => $e->getLine()));
>     watchdog('PHP', $message, array(), $e->severity);
>     $reset_date = FALSE;
>   }
> ```
> 
> ?
> 
> This way it gives nice and informative message like so:
> 
> ![image](https://user-images.githubusercontent.com/9372724/66720091-f4717e00-ee00-11e9-9f01-019642f5304d.png)

https://github.com/backdrop-contrib/demo/issues/1#issuecomment-541405783

**Comment by indigoxela** [comment]
Perhaps [watchdog_exception](https://api.backdropcms.org/api/backdrop/core%21includes%21bootstrap.inc/function/watchdog_exception/1) is a better choice here?

**Comment by indigoxela** [comment]
Revisiting this issue after some time, it occurs to me, we are talking about different things here.

@klonos suggested to add a utility to core to conveniently check for a valid timestamp - is that correct so far?

A **utility**, of course, shouldn't _ever_ throw an exception. Just silently check and give developers a true/false return.

Another note:   ` new DateTime('@' . $timestamp);` alone is way too permissive re the "timestamp".
An example: `'1234qwer'` is interpreted as 1970-01-01 00:20:34.
